# ENS210 - Group Project Team - (Batur,Banu,Yuşa)

This is a repository for `Spring Term 2018 - ENS210 Computational Biology - Group Project`. 

## Members

**Name:** [Batur Gültekin](https://www.linkedin.com/in/baturgultekin/)

**ID:** 24188

**E-mail:** baturgultekin@sabanciuniv.edu
__________

**Name:** [Banu Çetinkaya](20422)

**ID:** 20422

**E-mail:** banucetinkaya@sabanciuniv.edu
__________

**Name:** [Muhammed Yuşa Ergüven](19260)

**ID:** 19260

**E-mail:** merguven@sabanciuniv.edu


## Project Description (What to do?)

Characterize a protein family of your choice with respect to its evolutionary history, function, critical residues and domain(s).


## Daily Report Documentation

### 13 December 2018

- Proposal was written.

### 19 December 2018

- Initial 25 protein domains are selected.
- Conservation calculation score script was written.
- Initial conservation data is added.
- R script to generate conservation plot is initiated.

### 20 December 2018

- Protein domain data is updated - 37 more domains are added.
- Introduction and material method, first step of the project was documented.
- Alignment and plot is regenerated with new domains.

### 24 December 2018

- Protein domain data is changed => Old data is moved under Old-Datas-Scripts. New sequences added.
- New sequences alligned.
- New Tree is generated.
- Report is updated.
- `ENOG4111M9H.fa` is the sequences file.
- `ENOG4111M9H-aligned.fas` is the file contains the alligned sequences.
- `ENOG4111M9H-aligned-tree.nwk` is the Newick format of the tree generated.
- `exported-fig-tree.(png|jpg|pdf)` files are the Phylogenetic tree generated via Figtree.
- `calc-conserv-score.py` is the python script used to generate conservation values of Amino Acids.
- `identity.txt` is the result file generated by `calc-conserv-score.py`.
- `consensusTPD52.fa` is the consensus sequence.
- `identity_graph_plots.R` and `Non-gap-vs-identity-graph.R` are R scripts that are used to generate figures in the report, and `conserved_area_identity_TPD52.pdf` and `conserved_area_identity_and_gaps_for__each_position_TPD52.pdf` are results of the scripts.
- `gap-calc.py` is the Python script to calculate gaps at indexes of sequences, `numOfGaps.txt` is the result of the script.
- `TPD52_report.docx` is the Report of the project.



### Us during the whole project. (Not representational)

![](DOG.jpg)